### ISSUE TITLE
DataContractSerializer calls IDeserializationCallback before serialization is completed.

### ISSUE THREAD

**Issue by CodeLeopard** [dotnet/runtime#66788]
### Description

According to the documentation [here](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.ideserializationcallback?view=netstandard-2.0) the callback should be triggered when the entire object graph has been deserialized, however it is actually called after _that specific object_ has been deserialized.

### Reproduction Steps

Look at the output of this console application:
```csharp
using System;
using System.IO;
using System.Runtime.Serialization;

class Program
{
	[DataContract(IsReference = true)]
	class Foo : IDeserializationCallback
	{
		[DataMember]
		public Bar bar;

		[DataMember]
		public int sentinel;

		[OnDeserialized]
		public void OnDeserialized(StreamingContext context)
		{
			Console.WriteLine($"OnDeserialized   (attribute) Foo {sentinel,2}");
		}

		public void OnDeserialization(object sender)
		{
			Console.WriteLine($"OnDeserialization (callback) Foo {sentinel,2}");
		}
	}

	[DataContract(IsReference = true)]
	class Bar : IDeserializationCallback
	{
		[DataMember]
		public Foo foo;

		[OnDeserialized]
		public void OnDeserialized(StreamingContext context)
		{
			Console.WriteLine($"OnDeserialized   (attribute) Bar {foo.sentinel,2}");
		}

		public void OnDeserialization(object sender)
		{
			Console.WriteLine($"OnDeserialization (callback) Bar {foo.sentinel,2}");
		}
	}

	static void Main(string[] args)
	{
		var foo = new Foo();
		var bar = new Bar();

		foo.bar = bar;
		foo.sentinel = 42;
		bar.foo = foo;

		using var stream = new MemoryStream();
		var serializer = new DataContractSerializer(typeof(Foo));
		serializer.WriteObject(stream, foo);

		stream.Position = 0;

		var _ = (Foo)serializer.ReadObject(stream);
	}
}

```

### Expected behavior

Expected Console output:
```
OnDeserialized   (attribute) Bar  0
OnDeserialized   (attribute) Foo 42
OnDeserialization (callback) Foo 42
OnDeserialization (callback) Bar 42
```

The callback is issued _after_ the entire object graph is deserialized: so after all invocations of the method with `[OnDeserialized]` attribute.
Bar observes a value of 42 in the callback.

### Actual behavior

Actual Console output:
```
OnDeserialization (callback) Bar  0
OnDeserialized   (attribute) Bar  0
OnDeserialization (callback) Foo 42
OnDeserialized   (attribute) Foo 42
```

The callback is issued _before_ the entire object graph is deserialized: so before the last call to the methods with `[OnDeserialized]`.
Bar observes a value of 0 in the callback.

### Regression?

No, from what I can see in the history of the lines I reference below in 'Other information' it was initially implemented this way in [this pull request](https://github.com/dotnet/corefx/pull/13760) back when the code still lived in the corefx repository.

### Configuration

NetCoreApp3.1
Windows 10 Pro 21H2
x64 AMD

I originally encountered the issue in a class library project based on NetStandard2.0.

### Other information

I found that the [XmlFormatReaderGenerator](https://github.com/dotnet/runtime/blob/ac20efd2a26202d272604cb353d0d365a3f69c2d/src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs#L122-L125) adds a call to OnDeserialization immediately after reading the xml element.
Similarly [SurrogateDataContract](https://github.com/dotnet/runtime/blob/ac20efd2a26202d272604cb353d0d365a3f69c2d/src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/SurrogateDataContract.cs#L78-L79) also calls it immediately after reading the xml element.

While investigating the issue in my code I noted that the BinaryFormatter does trigger the callback at the expected time. I found a reference to `IDeserializzxationCallback` in [Objectmanager](https://github.com/dotnet/runtime/blob/a1bc0f34fc8ad77c31a1682841d92dbb20e39dd8/src/libraries/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/ObjectManager.cs#L686-L691) where the callback is registered to be called later.

Without the callback working as documented I don't know of a way for Bar to properly setup itself if it also depends on the state of Foo without me manually and explicitly propagating some trigger after the graph has been returned from the serializer.

**Comment by HongGit** [comment]
@CodeLeopard is this a regression from .NET framework?

**Comment by CodeLeopard** [comment]
@HongGit No: I just retested the reproduction example code with these TargetFrameworks: `net472;net48;netcoreapp3.1;net5.0;net6.0` and all of them produce identical (wrong) output.

**Comment by StephenMolloy** [comment]
`BinaryFormatter` and `DataContractSerializer` are two different serializers. More different than same. Sometimes sharing concepts between the two makes sense and sometimes it doesn't. The grouping of serializers/formatter in `System.Runtime` is loose and messy at times. Of course, now that `BinaryFormatter` has been retired, we're just left with a mish-mosh of concepts that might not make complete sense anymore without the context of BF around to explain them.

Anyway, with that said, since this has been the behavior for DCS going back into the days of .Net 4.X, I think it makes more sense to update the documentation than to change the behavior of DCS under peoples noses.
